ibm xl/16.1.1 compilation errors in nightly builds

Two separate build errors are occurring with the ibm/xl/16.1.1 compiler on Power8 architectures:

1. Older error: Undefined references leading to linking error following merge of PR #3719 

```
CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Other.cpp.o:(.toc+0x20): undefined reference to `Kokkos::Impl::ViewOffset<Kokkos::Impl::ViewDimension<0ul, 0ul, 0ul>, Kokkos::Experimental::LayoutTiled<(Kokkos::Iterate)1, (Kokkos::Iterate)2, 2u, 4u, 4u, 0u, 0u, 0u, 0u, 0u, true>, void>::SHIFT_2'
CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Other.cpp.o:(.toc+0x28): undefined reference to `Kokkos::Impl::ViewOffset<Kokkos::Impl::ViewDimension<0ul, 0ul, 0ul>, Kokkos::Experimental::LayoutTiled<(Kokkos::Iterate)2, (Kokkos::Iterate)2, 2u, 4u, 4u, 0u, 0u, 0u, 0u, 0u, true>, void>::SHIFT_2'
CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Other.cpp.o:(.toc+0x30): undefined reference to `Kokkos::Impl::ViewOffset<Kokkos::Impl::ViewDimension<0ul, 0ul, 0ul, 0ul>, Kokkos::Experimental::LayoutTiled<(Kokkos::Iterate)1, (Kokkos::Iterate)2, 2u, 4u, 4u, 2u, 0u, 0u, 0u, 0u, true>, void>::SHIFT_3'
CMakeFiles/KokkosCore_UnitTest_Serial1.dir/serial/TestSerial_Other.cpp.o:(.toc+0x38): undefined reference to `Kokkos::Impl::ViewOffset<Kokkos::Impl::ViewDimension<0ul, 0ul, 0ul, 0ul>, Kokkos::Experimental::LayoutTiled<(Kokkos::Iterate)2, (Kokkos::Iterate)2, 2u, 4u, 4u, 2u, 0u, 0u, 0u, 0u, true>, void>::SHIFT_3'
```

There are some apparent compiler issues with the replacement of `enum` with `static constexpr` usage, reversion of this change allowed an older SHA 08ba1104 of Kokkos to successfully compile for me

https://github.com/kokkos/kokkos/pull/3719/files#diff-b7c49ac8dda60407c7f5a565cd4fa8d696a6ed2dc33ad9f1e07d6807be254ca9 

@masterleinad do you have any suggestions or alternative guidance to reverting the changes above to work around the issue? This failure crept in right after I began OOO for several weeks, sorry for the delayed reporting of the issue


2. More recent changes in PR #3829 resulted in compilation errors beginning with the following:

```
01:06:33 /home/jenkins/white/workspace/Kokkos_develop_white_xl_1611_serial/kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp:94:40: error: template argument for template template parameter must be a class template or type alias template
01:06:33       typename T::template base_traits<msvc_workaround_get_next_base_traits,
01:06:33                                        ^
01:06:33 /home/jenkins/white/workspace/Kokkos_develop_white_xl_1611_serial/kokkos/core/src/impl/Kokkos_Utilities.hpp:177:45: note: in instantiation of template class 'Kokkos::Impl::msvc_workaround_get_next_base_traits<Kokkos::Impl::ExecutionSpaceTrait>' requested here
01:06:33 struct linearize_bases<GetBase, T, Ts...> : GetBase<T>::template apply<Ts...> {
01:06:33                                             ^
01:06:33 /home/jenkins/white/workspace/Kokkos_develop_white_xl_1611_serial/kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp:102:7: note: in instantiation of template class 'Kokkos::Impl::linearize_bases<msvc_workaround_get_next_base_traits, Kokkos::Impl::ExecutionSpaceTrait, Kokkos::Impl::GraphKernelTrait, Kokkos::Impl::IndexTypeTrait, Kokkos::Impl::IterationPatternTrait, Kokkos::Impl::LaunchBoundsTrait, Kokkos::Impl::OccupancyControlTrait, Kokkos::Impl::ScheduleTrait, Kokkos::Impl::WorkItemPropertyTrait, Kokkos::Impl::WorkTagTrait>' requested here
01:06:33     : linearize_bases<msvc_workaround_get_next_base_traits,
01:06:33       ^
01:06:33 /home/jenkins/white/workspace/Kokkos_develop_white_xl_1611_serial/kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp:107:7: note: in instantiation of template class 'Kokkos::Impl::AnalyzeExecPolicyBaseTraits<Kokkos::Impl::type_list<Kokkos::Impl::ExecutionSpaceTrait, Kokkos::Impl::GraphKernelTrait, Kokkos::Impl::IndexTypeTrait, Kokkos::Impl::IterationPatternTrait, Kokkos::Impl::LaunchBoundsTrait, Kokkos::Impl::OccupancyControlTrait, Kokkos::Impl::ScheduleTrait, Kokkos::Impl::WorkItemPropertyTrait, Kokkos::Impl::WorkTagTrait> >' requested here
01:06:33     : AnalyzeExecPolicyBaseTraits<execution_policy_trait_specifications> {
```

@dhollman any suggestions or guidance for things to try to resolve the error message above?


Reproducer instructions (White testbed, serial build)
```
module load cmake/3.18.0 ibm/xl/16.1.1 gcc/7.2.0

# Generate makefile script, extra flag magic for c++14 support
$KOKKOS_PATH/generate_makefile.bash --with-devices=Serial --arch=Power8 --compiler=/home/projects/ppc64le/ibm/xlC/16.1.1/bin/xlC --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized -isystem /home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu -qgcc_cpp_stdinc=/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0 -qxflag=disable__cplusplusOverride -I/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu" --cxxstandard="14" --kokkos-path=$KOKKOS_PATH

# Corresponding cmake line
cmake -DCMAKE_CXX_COMPILER=/home/projects/ppc64le/ibm/xlC/16.1.1/bin/xlC -DCMAKE_CXX_FLAGS="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized -isystem /home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu -qgcc_cpp_stdinc=/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0 -qxflag=disable__cplusplusOverride -I/home/projects/ppc64le/gcc/7.2.0/include/c++/7.2.0/powerpc64le-unknown-linux-gnu" -DKokkos_ENABLE_SERIAL=ON -DKokkos_ARCH_POWER8=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_CXX_STANDARD=14 -DCMAKE_BUILD_TYPE=RELEASE $KOKKOS_PATH
```

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -3,4 +3,4 @@ image:
 clone_folder: c:\projects\source
 build_script:
 - cmd: >-
-    cmake c:\projects\source -DKokkos_ENABLE_TESTS=ON -DCMAKE_CXX_FLAGS="/W0 /EHsc" -DCTEST_ARGS="-C Debug -V --output-on-failure" -DBUILD_NAME=MSVC-2019 -DBUILD_TYPE=Debug -DSITE=AppVeyor -DTARGET=install -P cmake/KokkosCI.cmake
+    cmake c:\projects\source -DKokkos_ENABLE_TESTS=ON -DCMAKE_CXX_FLAGS="/W0 /EHsc /d1reportClassLayoutChanges" -DCTEST_ARGS="-C Debug -V --output-on-failure" -DBUILD_NAME=MSVC-2019 -DBUILD_TYPE=Debug -DSITE=AppVeyor -DTARGET=install -P cmake/KokkosCI.cmake

--- a/core/src/impl/Kokkos_AnalyzePolicy.hpp
+++ b/core/src/impl/Kokkos_AnalyzePolicy.hpp
@@ -83,24 +83,12 @@ struct AnalyzeExecPolicy<void, void, Traits...>
 //------------------------------------------------------------------------------
 // Mix in the defaults (base_traits) for the traits that aren't yet handled
 
-// MSVC workaround: inheriting from more than one base_traits causes EBO to no
-// longer work, so we need to linearize the inheritance hierarchy
-template <class>
-struct msvc_workaround_get_next_base_traits;
-template <class T>
-struct msvc_workaround_get_next_base_traits {
-  template <class... Ts>
-  using apply =
-      typename T::template base_traits<msvc_workaround_get_next_base_traits,
-                                       Ts...>;
-};
-
 template <class TraitSpecList>
 struct AnalyzeExecPolicyBaseTraits;
 template <class... TraitSpecifications>
-struct AnalyzeExecPolicyBaseTraits<type_list<TraitSpecifications...>>
-    : linearize_bases<msvc_workaround_get_next_base_traits,
-                      TraitSpecifications...> {};
+struct KOKKOS_IMPL_ENFORCE_EMPTY_BASE_OPTIMIZATION
+    AnalyzeExecPolicyBaseTraits<type_list<TraitSpecifications...>>
+    : TraitSpecifications::base_traits... {};
 
 template <>
 struct AnalyzeExecPolicy<void>

--- a/core/src/impl/Kokkos_AnalyzePolicy.hpp
+++ b/core/src/impl/Kokkos_AnalyzePolicy.hpp
@@ -84,7 +84,7 @@ struct AnalyzeExecPolicy<void, void, Traits...>
 // Mix in the defaults (base_traits) for the traits that aren't yet handled
 
 template <class TraitSpecList>
-struct AnalyzeExecPolicyBaseTraits;
+struct KOKKOS_IMPL_ENFORCE_EMPTY_BASE_OPTIMIZATION AnalyzeExecPolicyBaseTraits;
 template <class... TraitSpecifications>
 struct KOKKOS_IMPL_ENFORCE_EMPTY_BASE_OPTIMIZATION
     AnalyzeExecPolicyBaseTraits<type_list<TraitSpecifications...>>

--- a/core/src/impl/Kokkos_Utilities.hpp
+++ b/core/src/impl/Kokkos_Utilities.hpp
@@ -161,27 +161,6 @@ struct type_list;
 // </editor-fold> end type_list }}}1
 //==============================================================================
 
-//==============================================================================
-// <editor-fold desc="MSVC linearize base workaround"> {{{1
-
-// MSVC workaround: more than two base classes (and more than one if the
-// hierarchy gets too deep) causes problems with EBO, so we need to linearize
-// the inheritance hierarchy to avoid losing EBO and ending up with an object
-// representation that is larger than it needs to be.
-// Note: by convention, the nested template in a higher-order metafunction like
-// GetBase is named apply, so we use that name here (this convention grew out
-// of Boost MPL)
-template <template <class> class GetBase, class...>
-struct linearize_bases;
-template <template <class> class GetBase, class T, class... Ts>
-struct linearize_bases<GetBase, T, Ts...> : GetBase<T>::template apply<Ts...> {
-};
-template <template <class> class GetBase>
-struct linearize_bases<GetBase> {};
-
-// </editor-fold> end MSVC linearize base workaround }}}1
-//==============================================================================
-
 }  // namespace Impl
 }  // namespace Kokkos
 

--- a/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp
+++ b/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp
@@ -57,9 +57,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct ExecutionSpaceTrait : TraitSpecificationBase<ExecutionSpaceTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     static constexpr auto execution_space_is_defaulted = true;
 
     using execution_space = Kokkos::DefaultExecutionSpace;

--- a/core/src/traits/Kokkos_GraphKernelTrait.hpp
+++ b/core/src/traits/Kokkos_GraphKernelTrait.hpp
@@ -58,9 +58,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct GraphKernelTrait : TraitSpecificationBase<GraphKernelTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     using is_graph_kernel = std::false_type;
   };
   template <class T>

--- a/core/src/traits/Kokkos_IndexTypeTrait.hpp
+++ b/core/src/traits/Kokkos_IndexTypeTrait.hpp
@@ -57,9 +57,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct IndexTypeTrait : TraitSpecificationBase<IndexTypeTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     static constexpr bool index_type_is_defaulted = true;
     using index_type = dependent_policy_trait_default;
   };

--- a/core/src/traits/Kokkos_IterationPatternTrait.hpp
+++ b/core/src/traits/Kokkos_IterationPatternTrait.hpp
@@ -55,9 +55,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct IterationPatternTrait : TraitSpecificationBase<IterationPatternTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     using iteration_pattern = void;  // TODO set default iteration pattern
   };
   template <class T>

--- a/core/src/traits/Kokkos_LaunchBoundsTrait.hpp
+++ b/core/src/traits/Kokkos_LaunchBoundsTrait.hpp
@@ -57,9 +57,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct LaunchBoundsTrait : TraitSpecificationBase<LaunchBoundsTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     static constexpr bool launch_bounds_is_defaulted = true;
 
     using launch_bounds = LaunchBounds<>;

--- a/core/src/traits/Kokkos_OccupancyControlTrait.hpp
+++ b/core/src/traits/Kokkos_OccupancyControlTrait.hpp
@@ -86,9 +86,7 @@ namespace Impl {
 // <editor-fold desc="Occupancy control trait specification"> {{{1
 
 struct OccupancyControlTrait : TraitSpecificationBase<OccupancyControlTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     using occupancy_control = Kokkos::Experimental::MaximizeOccupancy;
     static constexpr bool experimental_contains_desired_occupancy = false;
     // Default access occupancy_control, for when it is the (stateless) default

--- a/core/src/traits/Kokkos_ScheduleTrait.hpp
+++ b/core/src/traits/Kokkos_ScheduleTrait.hpp
@@ -58,9 +58,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct ScheduleTrait : TraitSpecificationBase<ScheduleTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     static constexpr auto schedule_type_is_defaulted = true;
 
     using schedule_type = Schedule<Static>;

--- a/core/src/traits/Kokkos_WorkItemPropertyTrait.hpp
+++ b/core/src/traits/Kokkos_WorkItemPropertyTrait.hpp
@@ -57,9 +57,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct WorkItemPropertyTrait : TraitSpecificationBase<WorkItemPropertyTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     using work_item_property = Kokkos::Experimental::WorkItemProperty::None_t;
   };
   template <class T>

--- a/core/src/traits/Kokkos_WorkTagTrait.hpp
+++ b/core/src/traits/Kokkos_WorkTagTrait.hpp
@@ -57,9 +57,7 @@ namespace Impl {
 // <editor-fold desc="trait specification"> {{{1
 
 struct WorkTagTrait : TraitSpecificationBase<WorkTagTrait> {
-  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
-  template <template <class> class GetBase, class... OtherTraits>
-  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+  struct base_traits {
     using work_tag = void;
   };
 };

